p2p/background/kad_discovery_baseline_test.go

```diff
@@ -0,0 +1,128 @@
+
+	bootstrapAddr, err := multiaddr.NewMultiaddr(fmt.Sprintf("%s/p2p/%s", addr1, host1.ID().String()))
+	require.NoError(t, err)
+
+	addr2, host2, _ := setupHostAndDiscovery(t, ctx, defaults.DefaultP2PPort+1, bootstrapAddr)
+	addr3, host3, _ := setupHostAndDiscovery(t, ctx, defaults.DefaultP2PPort+2, bootstrapAddr)
+
+	expectedPeerIDs := []libp2pPeer.ID{host1.ID(), host2.ID(), host3.ID()}
+
+	// TECHDEBT: consider using `host.ConnManager().Notifee()` to avoid sleeping here
+	// delay assertions for 500ms
+	time.Sleep(time.Millisecond * 500)
+
+	// assert that host2 has host3 in its peerstore
+	host2DiscoveredAddrs := host2.Peerstore().Addrs(host3.ID())
+	require.Lenf(t, host2DiscoveredAddrs, 1, "did not discover host3")
+	require.Equalf(t, addr3.String(), host2DiscoveredAddrs[0].String(), "did not discover host3")
+	require.ElementsMatchf(t, expectedPeerIDs, host2.Peerstore().Peers(), "host2 peer IDs don't match")
+
+	// assert that host3 has host2 in its peerstore
+	host3DiscoveredHost2Addrs := host3.Peerstore().Addrs(host2.ID())
+	require.Lenf(t, host3DiscoveredHost2Addrs, 1, "host3 did not discover host2")
+	require.Equalf(t, addr2.String(), host3DiscoveredHost2Addrs[0].String(), "host3 did not discover host2")
+	require.ElementsMatchf(t, expectedPeerIDs, host3.Peerstore().Peers(), "host3 peer IDs don't match")
+
+	// add another peer to network...
+	addr4, host4, _ := setupHostAndDiscovery(t, ctx, defaults.DefaultP2PPort+3, bootstrapAddr)
+	expectedPeerIDs = append(expectedPeerIDs, host4.ID())
+
+	// TECHDEBT: consider using `host.ConnManager().Notifee()` to avoid sleeping here
+	time.Sleep(time.Millisecond * 500)
+
+	// new host discovers existing hosts...
+	host4DiscoveredHost2Addrs := host4.Peerstore().Addrs(host2.ID())
+	require.Lenf(t, host4DiscoveredHost2Addrs, 1, "host4 did not discover host2")
+	require.Equalf(t, addr2.String(), host4DiscoveredHost2Addrs[0].String(), "host4 did not discover host2")
+
+	host4DiscoveredHost3Addrs := host4.Peerstore().Addrs(host3.ID())
+	require.Lenf(t, host4DiscoveredHost3Addrs, 1, "host4 did not discover host3")
+	require.Equalf(t, addr3.String(), host4DiscoveredHost3Addrs[0].String(), "host4 did not discover host3")
+
+	// existing hosts discovers host host...
+	host2DiscoveredHost4Addrs := host2.Peerstore().Addrs(host4.ID())
+	require.Lenf(t, host2DiscoveredHost4Addrs, 1, "host2 did not discover host4")
+	require.Equalf(t, addr4.String(), host2DiscoveredHost4Addrs[0].String(), "host2 did not discover host4")
+
+	host3DiscoveredHost4Addrs := host3.Peerstore().Addrs(host4.ID())
+	require.Lenf(t, host3DiscoveredHost4Addrs, 1, "host3 did not discover host4")
+	require.Equalf(t, addr4.String(), host3DiscoveredHost4Addrs[0].String(), "host3 did not discover host4")
+
+	require.ElementsMatchf(t, expectedPeerIDs, host4.Peerstore().Peers(), "host4 peer IDs don't match")
+}
+
+//nolint:unparam // DHT must exist but is otherwise "unused" (i.e. its API)
```

Comment:
Would it be clearer if we just refacto to `setupHost` and call `setupDHT(t, ctx, host, bootstrapAddr)` in place in the test above?

We can group the host creation and make it explicit when we need the bootstrap addr or not.

Lmk if you can "visualize" the alternate (potentially clearer) implementation

Reply:
I see what you're getting at. May I suggest that we leave it as is for now and see what API requirements emerge as we consolidate testutils and get into peer discovery testing? My thinking is that we will definitely need access to the DHT object at that point and I suspect there will be sufficient opportunity to refactor and reuse some of these testutils by then as well.